server/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -309,20 +315,14 @@ private void tryAllHandlers(final RestRequest request, final RestChannel channel
                 if (handlers == null) {
                     handler = null;
                 } else {
-                    handler = handlers.getHandler(requestMethod);
+                    handler = handlers.getHandler(requestMethod, compatibleWithVersion);
                 }
                 if (handler == null) {
                   if (handleNoHandlerFound(rawPath, requestMethod, uri, channel)) {
                       return;
                   }
                 } else {
-                    if(handler.compatibilityRequired() == false //regular (not removed) handlers are always dispatched
```

Comment:
no longer need that logic. each handler is registered with a compatibleWith version. 
The compatible API registers with Version.7
the current version register with Version.Current
this version is being used in handlers.getHandler to get the right handler